nbclient/execute.py

```diff
@@ -520,57 +582,14 @@ def run_cell(self, cell, cell_index=0, store_history=False):
         # after exec_reply was obtained from shell_channel, leading to the
         # aforementioned dropped data.
 
-        # These two variables are used to track what still needs polling:
-        # more_output=true => continue to poll the iopub_channel
-        more_output = True
-        # polling_exec_reply=true => continue to poll the shell_channel
-        polling_exec_reply = True
-
-        while more_output or polling_exec_reply:
-            if polling_exec_reply:
-                if self._passed_deadline(deadline):
-                    self._handle_timeout(exec_timeout, cell)
-                    polling_exec_reply = False
-                    continue
-
-                # Avoid exceeding the execution timeout (deadline), but stop
-                # after at most 1s so we can poll output from iopub_channel.
-                timeout = self._timeout_with_deadline(1, deadline)
-                exec_reply = self._poll_for_reply(parent_msg_id, cell, timeout)
-                if exec_reply is not None:
-                    polling_exec_reply = False
-
-            if more_output:
-                try:
-                    timeout = self.iopub_timeout
-                    if polling_exec_reply:
-                        # Avoid exceeding the execution timeout (deadline) while
-                        # polling for output.
-                        timeout = self._timeout_with_deadline(timeout, deadline)
-                    msg = self.kc.iopub_channel.get_msg(timeout=timeout)
-                except Empty:
-                    if polling_exec_reply:
-                        # Still waiting for execution to finish so we expect that
-                        # output may not always be produced yet.
-                        continue
-
-                    if self.raise_on_iopub_timeout:
-                        raise CellTimeoutError.error_from_timeout_and_cell(
-                            "Timeout waiting for IOPub output", self.iopub_timeout, cell
-                        )
-                    else:
-                        self.log.warning("Timeout waiting for IOPub output")
-                        more_output = False
-                        continue
-            if msg['parent_header'].get('msg_id') != parent_msg_id:
-                # not an output from our execution
-                continue
-
-            try:
-                # Will raise CellExecutionComplete when completed
-                self.process_message(msg, cell, cell_index)
-            except CellExecutionComplete:
-                more_output = False
+        exec_timeout = self._get_timeout(cell)
+        task_poll_output_msg = asyncio.ensure_future(
+            self.poll_output_msg(parent_msg_id, cell, cell_index)
+        )
+        exec_reply = await self._poll_for_reply(parent_msg_id, cell, exec_timeout)
+        if not task_poll_output_msg.done():
+            task_poll_output_msg.cancel()
```

Comment:
Never knew about cancel, can this mean a message can get lost?

Reply:
I don't think so, because messages are queued anyway in zmq. And we create a new task polling for messages just after that.

Reply:
But can it cancel, when it's off the zmq queue, say, it's in the json parser. How does it cancel like Ctrl-c.

Reply:
It can't do anything between two `await`, so it depends on the implementation. For instance, in `zmq.asyncio.Socket.recv_multipart` we might await `zmq.asyncio.Socket.recv` several times and so cancelling in the middle of it can lead to losing a message, you're right.

Reply:
Another way that I tried out was to leave the first `poll_output_msg` task running (with no timeout) and launch another one with a timeout, and then `asyncio.wait(both_tasks, return_when=asyncio.FIRST_COMPLETED)`, but it doesn't handle exceptions well (there is a `return_when=asyncio.FIRST_EXCEPTION` for that).
I can't think of a better solution right now.